File: java/src/PrCompressWord.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class PrCompressWord {
    public static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine();
        System.out.println(solution(s));
    }

    public static int solution(String s) {
        List<Character> str = new ArrayList<>();
        for (char c : s.toCharArray()) {
            str.add(c);
        }
        List<List<String>> wordList = new ArrayList<>();
        for (int length = 1; length < str.size(); length++) {
            List<String> list = new ArrayList<>();
            for (int idx = 0; idx < str.size(); idx += length) {
                if (idx + length < str.size()) {
                    list.add(s.substring(idx, idx + length));
                } else {
                    list.add(s.substring(idx));
                }
            }
            wordList.add(list);
        }
        int ans = s.length() == 1 ? 1 : Integer.MAX_VALUE;
        for (List<String> word : wordList) {
            int temp = 0;
            int cnt = 1;
            for (int i = 0; i < word.size() - 1; i++) {
                if (!word.get(i).equals(word.get(i + 1))) {
                    temp += cnt == 1 ? word.get(i).length() : word.get(i).length() + Integer.toString(cnt).length();
                    cnt = 1;
                } else {
                    cnt++;
                }
                if (i == word.size() - 2) {
                    temp += cnt == 1 ? word.get(i + 1).length() : word.get(i + 1).length() + Integer.toString(cnt).length();
                }
            }
            ans = Math.min(ans, temp);
        }
        return ans;
    }
}

File: kotlin/src/PrCompressWord.kt
import java.io.BufferedReader
import java.io.InputStreamReader

//https://school.programmers.co.kr/learn/courses/30/lessons/60057
class PrCompressWord {
    companion object{
        fun solve(){
            var br = BufferedReader(InputStreamReader(System.`in`))
            val s = br.readLine()
            fun solution(s: String): Int {
                val str = s.toList()
                val wordList = (1 until str.size).map { length ->
                    (0 until str.size step length).map { idx ->
                        if (idx + length < str.size) {
                            str.subList(idx, idx + length).joinToString("")
                        } else {
                            str.subList(idx, str.size).joinToString("")
                        }
                    }
                }
                var ans = if (s.length == 1) 1 else Int.MAX_VALUE

                wordList.map { word ->
                    var temp = 0
                    var cnt = 1
                    for (i in 0..word.size - 2) {
                        if (word[i] != word[i + 1]) {
                            temp += if (cnt == 1) word[i].length else word[i].length + cnt.toString().length
                            cnt = 1
                        } else {
                            cnt += 1
                        }
                        if (i == word.size - 2) {
                            temp += if (cnt == 1) word[i + 1].length else word[i + 1].length + cnt.toString().length
                        }
                    }
                    ans = Math.min(ans, temp)
                }
                return (ans)
            }
            println(solution(s))
        }
    }
}